### PR TITLE
Car collision visualization c img

### PR DESCRIPTION
**Pull request by ahmed-nad**

### CODE REVIEW

File: Car_Collision_Visulaization/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+
+add_library(
+    visualizer
+    src/CImg.h 
+    src/data_visualization.h
+    src/data_visualization.cpp
+    src/car_collision.h
+    src/car_collision.cpp
+)
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/CMakeLists.txt]: 1) Library names should start with capital letters
2) Visualizer library should only contain visualization code (car_collision.h and car_collision.cpp) should not be included as they are not part of the same functionality

So Visulization library should look like
add_library(
    Visualizer
    src/CImg.h 
    src/visualizer.h
    src/visualizer.cpp
    )

File: Car_Collision_Visulaization/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
+cmake_minimum_required(VERSION 3.0.0)
+project(DataVisualization VERSION 0.1.0 LANGUAGES C CXX)
+
+
+add_library(
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/CMakeLists.txt:5]: add_library(
    Car
    src/car.h
    src/car.cpp
    )
    
    We should have meaningful names and they are separate in functionality so they should be separate libraries

File: Car_Collision_Visulaization/src/car_collision.cpp [hunk 3]
```diff
@@ -1,5 +1,13 @@
+
 #include "car_collision.h"
 
+    Car::Car()
+    {
+        x_        = 0;
+        y_        = 0;
+        velocity_ = 0;
+        angle_    = 0;  
+    }
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/car_collision.cpp]: We don't need this constructor please delete it

File: Car_Collision_Visulaization/src/car_collision.h [hunk 4]
```diff
@@ -9,8 +12,11 @@ class Car
     public:
     double x_, y_ ;         
     double velocity_,angle_;
-
+    Car();
     explicit Car(double, double, double, double);
+    void move();
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/car_collision.h]: please define variables and functions separately
public: 
  double x_;
  double y_;
  double velocity_;
  
public:
  explicit Car (double x, double y, double velocity, double angle);
  void move();
  bool isColliding(const Car& target_vehicle);

File: Car_Collision_Visulaization/src/car_collision.h [hunk 5]
```diff
@@ -9,8 +12,11 @@ class Car
...
 };
 
-bool isCalculatingCollision(const Car&, const Car&);
\ No newline at end of file
+bool isCalculatingCollision(const Car&, const Car&);
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/car_collision.h]: This should be is_colliding() what you have written does not make any sense

File: Car_Collision_Visulaization/src/data_visualization.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
+
+
+#include "data_visualization.h"
+
+
+Visualizer :: Visualizer()
+{
+    length_ = 0;
+    width_  = 0;
+}
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/data_visualization.cpp]: Delete this one. And make the other one explicit  as we can not have a visualizer without a window size

File: Car_Collision_Visulaization/src/data_visualization.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    unsigned char blue[3] = {0, 0, 255};
+    drawRectangleShape(canvas, car2, blue);
+
+    CImgDisplay display(canvas, "Rectangle and Circle"); 
+
+    while (!display.is_closed()) 
+    {
+        display.wait();
+    }
+}
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/data_visualization.cpp]: This is wrong design. This way you will only get the initial values of cars and than you won't be able to do anything else. 
1) The constructor of Visualizer should have these lines
   
CImg<unsigned char> canvas(length, width, 1, 3, 255); 
CImgDisplay display(canvas, "Rectangle and Circle"); 

2) update() should accept const std::vector<Car>& cars as we want be able to draw multiple cars.
3) Car should contain length and width of car as well. Which would be used in update() to draw vehicles.

void Visualizer :: update( const std::vector<Car>& all_cars)
{
   for (const Car& car : all_cars) 
   {
       drawRectangleShape(canvas, car, red);
    }
}

File: Car_Collision_Visulaization/src/data_visualization.h [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+#include "car_collision.h"
+using namespace cimg_library;
+
+class Visualizer 
+{
+    public:
+    size_t length_;
+    size_t width_;
+    Visualizer();
+    Visualizer(size_t length, size_t width);
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/data_visualization.h]: please update like i showed you in the other class

File: Car_Collision_Visulaization/src/car_collision.h (hunk 4 above)
Comment (ahmed-nad) [reply Car_Collision_Visulaization/src/car_collision.h]: Car should have length and width as well

File: Car_Collision_Visulaization/main.cpp [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+    Car car1(200, 2, 5, 45 );    //angle in radian    
+    Car car2(55 , 4, 7, 135);  
+    Visualizer visualize; 
+ 
+    // Car car1(2, 2, 5, 45 );    //angle in radian    
+    // Car car2(5, 4, 7, 135); 
+    // Visualizer visualize;
+
+    visualize.update(car1,car2);
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/main.cpp]: Initialize all cars in a vector .

std::vector<Car> all_cars;
all_cars.push_back(200,2,5,45);
all_cars.push_back(55,4,7,135);

while (true)
{ 
  for (const Car& car : all_cars)
    car.move();
   visualize.update(all_cars)
}

**Review by ahmed-nad — COMMENTED** [review]
A lot of improvements required

File: Car_Collision_Visulaization/src/car_collision.cpp (hunk 3 above)
Comment (Israr4) [reply Car_Collision_Visulaization/src/car_collision.cpp]: I need that default constructor to be used in code,if i don,t write a default constructor then compiler gives error.

File: Car_Collision_Visulaization/main.cpp (hunk 9 above)
Comment (Israr4) [reply Car_Collision_Visulaization/main.cpp]: visualize object needs default constructor there

File: Car_Collision_Visulaization/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    src/CImg.h 
+    src/data_visualization.h
+    src/data_visualization.cpp
+)
+
+add_library(
+    Car
+    src/car_collision.h
+    src/car_collision.cpp
+)
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/CMakeLists.txt]: car_collision.h and car_collision.cpp
are not meaningful names anymore please change it. As i suggested before please change it to car.h and car.cpp

File: Car_Collision_Visulaization/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,26 @@
+cmake_minimum_required(VERSION 3.0.0)
+project(DataVisualization VERSION 0.1.0 LANGUAGES C CXX)
+
+
+add_library(
+    Visualizer
+    src/CImg.h 
+    src/data_visualization.h
+    src/data_visualization.cpp
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/CMakeLists.txt]: data_visualization.h and data_visualization.cpp are not meaningful names please change them 
as i suggested before should be 
visualizer.h and visualizer.cpp

File: Car_Collision_Visulaization/src/data_visualization.h [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public:
+    size_t length_;
+    size_t width_;
+    Visualizer();
+
+    public:
+    explicit Visualizer(size_t length, size_t width);
+    void drawRectangleShape(Car car, unsigned char color[3]);
+    void update(std::vector<Car> &all_cars);
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/data_visualization.h]: delete Visualizer() please we shouldn't need it

File: Car_Collision_Visulaization/src/data_visualization.cpp [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+    for (int i = 0; i < 10; ++i) 
+    {
+        std::cout<<"Hello"<<std::endl;
+        canvas.fill(100);
+        car.move();
+        canvas.draw_rectangle(car.x_ - (carWidth / 2), car.y_ - (carHeight / 2), car.x_ + (carWidth / 2), car.y_ + (carHeight / 2), color, 1.0f);
+        canvas.display(display);
+        cimg::wait(100);
+    }
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/data_visualization.cpp:34]: Why the loop here?

File: Car_Collision_Visulaization/src/data_visualization.cpp [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+Visualizer :: Visualizer(size_t length, size_t width)
+{
+    length_ = length;
+    width_  = width;
+}
+
+void Visualizer :: drawRectangleShape( Car car, unsigned char color[3]) 
+{
+    int carWidth  = 50;
+    int carHeight = 30;
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/data_visualization.cpp]: This should be inside each car object

File: Car_Collision_Visulaization/src/data_visualization.cpp [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    int length = 1000;
+    int width  = 1000;
+    Visualizer(length,width);
+
+    CImg<unsigned char> canvas(length, width, 1, 3, 255); 
+    CImgDisplay display(canvas, "Rectangle and Circle");
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/data_visualization.cpp]: This should be initialized once in the constructor

File: Car_Collision_Visulaization/src/data_visualization.cpp (hunk 13 above)
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/data_visualization.cpp]: Why put the for loop here? we don't need it.

**Review by ahmed-nad — COMMENTED** [review]
Ye design theek nai hai. Ye socho ke hr class ne kea krna hai.
1) Car ki jo class hai usmein uski length aur width honi chaiye
2) Visualizer ki class sirf canvas aur baki cheezo ko initialize kregi aur uske update() function mein jo kuch ayega usko wo draw krdegi. While loop uske andhr nai hona chaiye. Jaha se call horha while loop waha hona chaiye. Jese meine pehle bola tha ke Canvas object tm visualizer mein initialize kro jo length, width ati constructor mein uske through. Phr us canvas object ko update() function mein reuse kro.

File: Car_Collision_Visulaization/src/visualizer.cpp [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+Visualizer :: Visualizer(size_t length, size_t width):canvas_(length, width, 1, 3, 255), display_(canvas_, "Rectangle and Circle")
+{
+    length_  = length;
+    width_   = width;  
+}
+
+void Visualizer :: drawRectangleShape(CImg<unsigned char>& canvas, Car car, unsigned char color[3])  
+{
+    std::cout<<"Hello"<<std::endl;
```
Comment (ahmed-nad) [inline Car_Collision_Visulaization/src/visualizer.cpp]: remove

Comment (Israr4) [reply Car_Collision_Visulaization/src/visualizer.cpp]: removed

**Pull request merged**